### PR TITLE
Create a new plugin supports /trick-or-treat

### PR DESCRIPTION
**Pull request by chaodaiG**
Collaborated with mpherman2@ on this project, and figured it might bring some fun to k8s developers as we approach halloween

/cc @ mpherman2 @michelle192837 @e-blackwelder

### CODE REVIEW

File: prow/plugins/trickortreat/trickortreat.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var (
+	match        = regexp.MustCompile(`(?mi)^/trick(\-)?or(\-)?treat(?: (.+))?\s*$`)
+	trickortreat = &realClowder{}
+
+	// A curated list that contains safe images URLs that we feel confident to share.
+	// TODO: use an API for obtaining list of images, for example from https://www.pexels.com/api/documentation.
+	candiesImgs = []string{
```
Comment (MushuEE) [inline prow/plugins/trickortreat/trickortreat.go:40]: No tricks? https://i.redd.it/snqk3beim6x31.jpg

Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go:40]: Initially we were thinking about `/trick-or-treat <CANDY NAME>` to be more fun, but due to the limitation of manually curated list, we decided not to go that far :)

File: prow/plugins/trickortreat/trickortreat.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	pluginName = "trickortreat"
+)
+
+func init() {
+	plugins.RegisterGenericCommentHandler(pluginName, handleGenericComment, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, _ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The trickortreat plugin adds a trickortreat image to an issue or PR in response to the `/trick-or-treat` command.",
```
Comment (michelle192837) [inline prow/plugins/trickortreat/trickortreat.go:112]: nit: Should this also use 'candy' like below?

**Comment by michelle192837** [comment]
Broadly looks good, added a couple of comments!

File: prow/plugins/trickortreat/trickortreat.go (hunk 2 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go]: Good catch, fixed

File: prow/plugins/trickortreat/trickortreat.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var (
+	match        = regexp.MustCompile(`(?mi)^/trick(\-)?or(\-)?treat(?: (.+))?\s*$`)
+	trickortreat = &realClowder{}
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go]: fixed

File: prow/plugins/trickortreat/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+api-key
```
Comment (fejta) [inline prow/plugins/trickortreat/.gitignore:1]: Why do we need this if we aren't calling an API?

File: prow/plugins/trickortreat/trickortreat.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var (
+	match        = regexp.MustCompile(`(?mi)^/trick(\-)?or(\-)?treat(?: (.+))?\s*$`)
+	trickortreat = &realSnicker{}
+
+	// A curated list that contains safe images URLs that we feel confident to share.
+	// TODO: use an API for obtaining list of images, for example from https://www.pexels.com/api/documentation.
+	candiesImgs = []string{
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:40]: Will you please at least put these in another file?

File: prow/plugins/trickortreat/trickortreat.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+	org := e.Repo.Owner.Login
+	repo := e.Repo.Name
+	number := e.Number
+
+	for i := 0; i < 3; i++ {
+		imgURL, err := c.readImage()
+		if err != nil {
+			log.WithError(err).Error("Failed to get img")
+			continue
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:184]: We should sleep before continuing

File: prow/plugins/trickortreat/trickortreat_test.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+				break
+			}
+		})
+	}
+}
+
+func TestReadImage(t *testing.T) {
+	if img, err := trickortreat.readImage(); err != nil {
+		t.Errorf("Could not read candies from %#v: %v", trickortreat, err)
+	} else {
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat_test.go:64]:
```golang
if err := foo(); err != nil {
  t.Errorf("whatever")
  return
}
t.Log(img)
```

is the preferred style

https://dave.cheney.net/practical-go/presentations/qcon-china.html#_return_early_rather_than_nesting_deeply

File: prow/plugins/trickortreat/trickortreat_test.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+		}
+		err := handle(fc, logrus.WithField("plugin", pluginName), e, &fakeSnicker{tc.readImgErrString})
+		if !tc.shouldError && err != nil {
+			t.Errorf("%s: didn't expect error: %v", tc.name, err)
+			continue
+		} else if tc.shouldError && err == nil {
+			t.Errorf("%s: expected an error to occur", tc.name)
+			continue
+		}
+		if tc.shouldComment && len(fc.IssueComments[5]) != 1 {
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat_test.go:151]: else if

File: prow/plugins/trickortreat/trickortreat_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	"testing"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/github/fakegithub"
+)
+
+type fakeSnicker struct {
+	errStr string
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat_test.go:32]: `err error` seems more natural and flexible here than a string

File: prow/plugins/trickortreat/trickortreat.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+func handle(gc githubClient, log *logrus.Entry, e *github.GenericCommentEvent, c snicker) error {
+	// Only consider new comments.
+	if e.Action != github.GenericCommentActionCreated {
+		return nil
+	}
+	// Make sure they are requesting a cat
+	mat := match.FindStringSubmatch(e.Body)
+	if mat == nil {
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:172]: `mat` isn't used after this, please combine this with the above line

File: prow/plugins/trickortreat/trickortreat.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+		imgIndex := rand.Intn(len(candiesImgs))
+		imgURL := candiesImgs[imgIndex]
+		// checking size, GitHub doesn't support big images
+		toobig, err = github.ImageTooBig(imgURL)
+		if err == nil && !toobig {
+			return imgURL, nil
+		}
+	}
+	if err == nil && toobig {
+		err = fmt.Errorf("Image %q too big", imgURL)
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:151]: `image` not `Image`

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Also, this means 3 images were too big and/or had other errors, right? Seems like we should either include all 3 URLs in the error or not do that here and log it in the `for` loop

File: prow/plugins/trickortreat/trickortreat.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+func (c *realSnicker) readImage() (string, error) {
+	var imgURL string
+	var err error
+	var toobig bool
+	for i := 0; i < 3; i++ {
+		imgIndex := rand.Intn(len(candiesImgs))
+		imgURL := candiesImgs[imgIndex]
+		// checking size, GitHub doesn't support big images
+		toobig, err = github.ImageTooBig(imgURL)
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:145]: `tooBig`

File: prow/plugins/trickortreat/trickortreat.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var (
+	match        = regexp.MustCompile(`(?mi)^/trick(\-)?or(\-)?treat(?: (.+))?\s*$`)
+	trickortreat = &realSnicker{}
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:36]: * `trickOrTreat`
* No reason for this to be a pointer (so `trickOrTreat realSnicker` should work just fine)

File: prow/plugins/trickortreat/trickortreat.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package trickortreat adds halloween images to an issue or PR in response to a /trick-or-treat comment
+package trickortreat
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:18]: consider `trick-or-treat` for the package and plugin name

Three unseparated words is IMO too many, (see merge-mmethod-comment, owners-label, etc plugins, external-plugins, pod-utils packages)

File: prow/plugins/trickortreat/trickortreat.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+type snicker interface {
+	readImage() (string, error)
+}
+
+type realSnicker struct {
+}
+
+func (c *realSnicker) readImage() (string, error) {
+	var imgURL string
```
Comment (fejta) [inline prow/plugins/trickortreat/trickortreat.go:138]: This is always nil outside the `for` loop

**Review by fejta — COMMENTED** [review]
/lint

File: prow/plugins/trickortreat/trickortreat_test.go (hunk 9 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat_test.go]: agree, done

File: prow/plugins/trickortreat/trickortreat_test.go (hunk 7 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat_test.go]: done, img is used so changed the first line to two lines

File: prow/plugins/trickortreat/trickortreat_test.go (hunk 8 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat_test.go]: done

File: prow/plugins/trickortreat/.gitignore (hunk 4 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/.gitignore]: great catch, deleted

File: prow/plugins/trickortreat/trickortreat.go (hunk 14 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go:18]: changed the plugin name. Go doesn't seem like dash in package name, so keep it as is

File: prow/plugins/trickortreat/trickortreat.go (hunk 13 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go]: Done

File: prow/plugins/trickortreat/trickortreat.go (hunk 5 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go]: Done

File: prow/plugins/trickortreat/trickortreat.go (hunk 15 above)
Comment (chaodaiG) [reply prow/plugins/trickortreat/trickortreat.go:75]: That's a good catch, fixed

**Review by fejta — APPROVED** [review]

**Pull request merged**